tmva/sofie/inc/TMVA/ROperator_Reduce.hxx

```diff
@@ -0,0 +1,301 @@
+            throw std::runtime_error("TMVA SOFIE Reduce Op Input Tensor " + fNX + " is not found in model");
+        }
+        fShapeX = model.GetTensorShape(fNX);
+        // find shape of Y and add it in the list of intermediate tensors
+        fShapeY = ShapeInference(fShapeX);
+        model.AddIntermediateTensor(fNY, model.GetTensorType(fNX), fShapeY);
+    }
+
+    std::string Generate(std::string OpName){
+      OpName = "op_" + OpName;
+      if (fShapeX.empty() || fShapeY.empty()) {
+         throw std::runtime_error("TMVA SOFIE Reduce Op called to Generate without being initialized first");
+      }
+
+      size_t outputLength = TMVA::Experimental::SOFIE::ConvertShapeToLength(fShapeY);
+
+      auto inputStrides = TMVA::Experimental::SOFIE::UTILITY::ComputeStrideFromShape(fShapeX);
+      auto outputStrides = TMVA::Experimental::SOFIE::UTILITY::ComputeStrideFromShape(fShapeY);
+      
+   // write here according to size of shape
+   // in generation code can be done automatically
+   // i0 =  i / s0 ; i1 = (i % s0) / s1 ; i2 = ( (i % s0) % s1 ) / s2 and so on
+   // and we have for the inverse
+   // i = i0 * s0 + i1 * s1 + i2 * s2 + i3 * s3 ....
+
+   // don't need to divide by last stride s[n-1] since it is 1 by definition
+      
+      std::stringstream out;
+      out << "\n//----  operator " << std::string(ReduceOperatorTrait<T,Op>::Name()) << "  " << OpName << "\n";
+      out << SP << "size_t dim = " << fShapeY.size() << ";\n";
+
+      out << SP << "std::vector<size_t> idx(dim);";
+
+      
+      out << SP << "for (size_t i = 0; i < " << outputLength << "; i++) {\n";
+      
+      // write here according to size of shape
+      // in generation code can be done automatically
+      // i0 =  i / s0 ; i1 = (i % s0) / s1 ; i2 = ( (i % s0) % s1 ) / s2 and so on
+      // and we have for the inverse
+      // i = i0 * s0 + i1 * s1 + i2 * s2 + i3 * s3 ....
+
+      // don't need to divide by last stride s[n-1] since it is 1 by definition
+      
+      out << SP << SP << "idx[j] = i;\n";
+      out << SP << SP << "size_t k = 0;\n";
+      out << SP << SP << SP << "for(k=0; k < dim-1; k++){\n";
+      out << SP << SP << SP << "idx[k] = idx[k] %" << outputStrides << "[k];\n";
+      out << SP << SP << SP << "};\n"; 
+      out << SP << SP << "idx[j] = idx[j] /" << outputStrides << "[k];\n";
+      
+
+      out << SP << "assert(idx[" << fAttrAxes << "] == 0);\n";  // we can avoid computing this for the reduction axis which by definition is always zero
+
+      out << SP << "float sum = 0;\n";
+      out << SP << SP << " for (size_t k = 0; k < inputShape[" << fAttrAxes << "]; k++) { \n";
+      out << SP << SP << "  idx[" << fAttrAxes << "] = k;\n";
+       // compute input index j 
+      out << SP << SP << "size_t l = 0;\n";
+      out << SP << SP << "size_t m = 0;\n";
+      out << SP << SP << SP << "for(m=0; m < dim-1; m++){\n";
+      out << SP << SP << SP << "l += idx[m] *" << inputStrides << "[m];\n";
```

Comment:
Also for inputStrides should be as for output strides. 

Otherwise instead of looping above on k and here on m you can in generated code just write the lines for each tensor dimension. I think this is better, because it avoids the loop and we profit better from the generated code.